server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -44,48 +49,125 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
 }
 
 object ParsedMultiPart {
-  def empty: ParsedMultiPart = new ParsedMultiPart(Map.empty, Array.empty)
+  def empty: ParsedMultiPart = new ParsedMultiPart(
+    Map.empty,
+    new ByteArrayInputStream(Array.empty)
+  )
 
   sealed trait ParseState
   case object Default extends ParseState
   case object AfterBoundary extends ParseState
+  case object AfterParsedBody extends ParseState
   case object AfterHeaderSpace extends ParseState
 
   private case class ParseData(
       currentPart: ParsedMultiPart,
       completedParts: List[ParsedMultiPart],
       parseState: ParseState
   ) {
+    def done(): ParseData = this
     def changeState(state: ParseState): ParseData = this.copy(parseState = state)
     def addHeader(header: String): ParseData = this.copy(currentPart = currentPart.addHeader(header))
-    def addBody(body: Array[Byte]): ParseData = this.copy(currentPart = currentPart.copy(body = currentPart.body ++ body))
-    def completePart(): ParseData = this.currentPart.getName match {
+    def completePart(body: InputStream): ParseData = this.currentPart.getName match {
       case Some(_) =>
         this.copy(
-          completedParts = completedParts :+ currentPart,
+          completedParts = completedParts :+ currentPart.copy(body = body),
           currentPart = empty,
-          parseState = AfterBoundary
+          parseState = AfterParsedBody
         )
-      case None => changeState(AfterBoundary)
+      case None => changeState(AfterParsedBody)
     }
   }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
+  private def toTruncatedInputStream(streamAndFile: PartStream, stopAt: Int): InputStream = streamAndFile match {
+    case ByteStream(stream) =>
+      val bytes = stream.toByteArray
+      new ByteArrayInputStream(bytes.slice(0, stopAt))
+    case FileStream(f, fileOutputStream) =>
+      fileOutputStream.getChannel.truncate(stopAt.toLong)
+      fileOutputStream.close()
+      new FileInputStream(f)
+  }
+
+  private def readUntilNewline(inputStream: InputStream): Array[Byte] =
     Iterator
-      .continually(reader.readLine())
```

Comment:
While this previous way felt more readable, i couldn't figure out a good way to "chunk" the data up when determining whether the threshold was passed or not when reading lines (What is the body doesn't have any newlines), so i think the new way is better(?)

Reply:
I think here a good-old for loop with two mutable vars for last two bytes + a `ByteArrayOutputStream` for accumulating the values would be much more performant